java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/connector/source/common/PostgresValidator.java

```diff
@@ -384,17 +388,91 @@ private void validatePublicationConfig(boolean isSuperUser) throws SQLException
-                            + pubName
-                            + "' to be created with `publish_via_partition_root = true`. \nHint: you can run `SELECT pubviaroot from pg_publication WHERE pubname = '"
-                            + pubName
-                            + "'` in the upstream Postgres to check.");
+
+        if (!partitions.isEmpty() && isPublicationExists) {
+            // `pubviaroot` in `pg_publication` is added after PG v13, before which PG does not
+            // allow adding partitioned table to a publication. So here, if partitions.isEmpty() is
+            // false, we can safely check the value of `pubviaroot` of the publication here.
+            boolean isPublicationViaRoot = false;
+            try (var stmt =
+                    jdbcConnection.prepareStatement(
+                            ValidatorUtils.getSql("postgres.publication_pubviaroot"))) {
+                stmt.setString(1, pubName);
+                var res = stmt.executeQuery();
+                if (res.next()) {
+                    isPublicationViaRoot = res.getBoolean(1);
+                }
+            }
+            if (!isPublicationViaRoot) {
+                // Make sure the publication are created with `publish_via_partition_root = true`,
+                // which is required by partitioned tables.
+                throw ValidatorUtils.invalidArgument(
+                        "Table '"
+                                + tableName
+                                + "' has partitions, which requires publication '"
+                                + pubName
+                                + "' to be created with `publish_via_partition_root = true`. \nHint: you can run `SELECT pubviaroot from pg_publication WHERE pubname = '"
+                                + pubName
+                                + "'` in the upstream Postgres to check.");
+            }
+        }
+        //  Only after v13, PG allows adding a partitioned table to a publication. So, if the
+        // version is before v13, the tables in a publication are always partition leaves, we don't
+        // check their ancestors and descendants anymore.
+        if (isPublicationExists && pgVersion >= 13) {
+            List<String> family = new ArrayList<>();
+            boolean findRoot = false;
+            String currentPartition = tableName;
+            while (!findRoot) {
+                try (var stmt =
+                        jdbcConnection.prepareStatement(
+                                ValidatorUtils.getSql("postgres.partition_parent"))) {
+                    stmt.setString(
+                            1, String.format("\"%s\".\"%s\"", this.schemaName, currentPartition));
+                    stmt.setString(
+                            2, String.format("\"%s\".\"%s\"", this.schemaName, currentPartition));
+                    stmt.setString(
+                            3, String.format("\"%s\".\"%s\"", this.schemaName, currentPartition));
```

Comment:
May use a variable to eliminate redundancy of `String.format("\"%s\".\"%s\"", this.schemaName, currentPartition)`

Reply:
Updated